Archive/Imo/Imo2024Q5.lean

```diff
@@ -0,0 +1,1050 @@
+import Mathlib.Data.Nat.Dist
+import Mathlib.Order.Fin.Basic
+import Mathlib.Tactic.IntervalCases
+
+/-!
+# IMO 2024 Q5
+
+Turbo the snail plays a game on a board with $2024$ rows and $2023$ columns. There are hidden
+monsters in $2022$ of the cells. Initially, Turbo does not know where any of the monsters are,
+but he knows that there is exactly one monster in each row except the first row and the last
+row, and that each column contains at most one monster.
+
+Turbo makes a series of attempts to go from the first row to the last row. On each attempt,
+he chooses to start on any cell in the first row, then repeatedly moves to an adjacent cell
+sharing a common side. (He is allowed to return to a previously visited cell.) If he reaches a
+cell with a monster, his attempt ends and he is transported back to the first row to start a
+new attempt. The monsters do not move, and Turbo remembers whether or not each cell he has
+visited contains a monster. If he reaches any cell in the last row, his attempt ends and the
+game is over.
+
+Determine the minimum value of $n$ for which Turbo has a strategy that guarantees reaching
+the last row on the $n$th attempt or earlier, regardless of the locations of the monsters.
+
+We follow the solution from the
+[official solutions](https://www.imo2024.uk/s/IMO2024-solutions-updated.pdf). To show that $n$
+is at least $3$, it is possible that the first cell Turbo encounters in the second row on his
+first attempt contains a monster, and also possible that the first cell Turbo encounters in the
+third row on his second attempt contains a monster. To show that $3$ attempts suffice, the first
+attempt can be used to locate the monster in the second row; if this is not at either side of
+the board, two more attempts suffice to pass behind that monster and from there go along its
+column to the last row, while if it is at one side of the board, the second attempt follows a
+zigzag path such that if it encounters a monster the third attempt can avoid all monsters.
+-/
+
+
+namespace Imo2024Q5
+
+/-! ### Definitions for setting up the problem -/
+
+-- There are N monsters, N+1 columns and N+2 rows.
+variable {N : ℕ}
+
+/-- A cell on the board for the game. -/
+abbrev Cell (N : ℕ) : Type := Fin (N + 2) × Fin (N + 1)
+
+/-- A row that is neither the first nor the last (and thus contains a monster). -/
+abbrev InteriorRow (N : ℕ) : Type := (Set.Icc 1 ⟨N, by omega⟩ : Set (Fin (N + 2)))
+
+/-- Data for valid positions of the monsters. -/
+abbrev MonsterData (N : ℕ) : Type := InteriorRow N ↪ Fin (N + 1)
+
+/-- The cells with monsters as a Set, given an injection from rows to columns. -/
+def MonsterData.monsterCells (m : MonsterData N) :
+    Set (Cell N) :=
+  Set.range (fun x : InteriorRow N ↦ ((x : Fin (N + 2)), m x))
+
+/-- Whether two cells are adjacent. -/
+def Adjacent (x y : Cell N) : Prop :=
+  Nat.dist x.1 y.1 + Nat.dist x.2 y.2 = 1
+
+/-- A valid path from the first to the last row. -/
+structure Path (N : ℕ) where
```

Comment:
Have you considered using `SimpleGraph.Walk` instead of defining your own `Path` like this? (I'm not asking you to rewrite this PR, because I expect that would be a lot of work; I'm just curious whether you had any thoughts about how possible it would be to use that existing API.)

https://github.com/leanprover-community/mathlib4/blob/fd92bc36215a9632f15c6482fb0132d24e6c49ab/Mathlib/Combinatorics/SimpleGraph/Walk.lean#L45-L56

Reply:
One advantage that I imagine `SimpleGraph.Walk` might have is that it could make it easier to construct walks by concatenating smaller walks. That might allow more compositional reasoning, compared to the `Path.ofFn` logic that you have here.

Reply:
I've thought about the question of when IMO problem formalizations should refer to graphs in the statement, yes, though I haven't written up any of my conclusions about what uniform formalization conventions for such problems might look like (and my thoughts on such questions evolved while I was writing formal statements for all the non-geometry 2024 shortlist problems, so some of those formalizations don't necessarily follow what I now think would be the best conventions - even apart from the differences between conventions for stating a problem on its own versus conventions for stating a problem when accompanied by a solution).

On the matter of graphs, my inclination is:
* if the problem explicitly refers to graphs (IMO 1991 P4) then of course use them in the formal statement;
* if the problem is essentially about graphs, just using different terminology rather than the word "graph" (e.g. IMO 2019 P3 or 2023 shortlist C7) then it's also appropriate to use them in the formal statement;
* but if the problem only incidentally involves some kind of adjacency relation, which could be expressed in terms of a graph but the problem isn't really a graph theory problem in shallow disguise (as with this problem), then it makes more sense to express it without graphs (of course a (human or AI) solver can translate into graphs for a solution if they wish);
* which is similar to the case of problems not even involving an adjacency relation (so obviously not expressed in terms of graphs) where graphs are nevertheless useful in solving the problem (e.g. IMO 2020 P3, which is convenient to solve using Eulerian circuits on multigraphs - not in mathlib at present, I think - but where nothing like that appears in the statement).

I don't, however, think that `SimpleGraph.Walk` would make things any simpler. I considered joining smaller pieces (via having a structure that only requires adjacency but doesn't constrain the end points), but defining such pieces would still require either a custom inductive definition to build a piece of length N (say), or use of `ofFn` (which is built on the inductive `Fin.foldr`) in defining the pieces, and anything involving composing multiple custom inductive definitions is likely to end up more complicated (with local API needed for each inductive definition) than just defining each path using `ofFn` applied to a function defined directly with `if` and then letting the automation of `split_ifs <;> simp <;> omega` deal with proving adjacency everywhere. And `List` has much more existing API to take advantage of in the statement and solution than `Walk` does.

You could of course try formalizing a non-graph-theory problem with adjacency (e.g. IMO 2022 P6 or IMO 2023 P5) in a way using mathlib graphs and see how it turns out; it's not what I think of as the most natural way to state such a problem, but such problems admit lots of different approaches to stating them formally.